src/BitStringAddresses/fockaddress.jl

```diff
@@ -0,0 +1,195 @@
+"""
+    AbstractFockAddress
+
+Supertype representing a Fock state.
+"""
+abstract type AbstractFockAddress{N,M} end
+
+"""
+    num_particles(::Type{<:AbstractFockAddress})
+    num_particles(::AbstractFockAddress)
+
+Number of particles represented by address.
+"""
+num_particles(a::AbstractFockAddress) = num_particles(typeof(a))
+num_particles(::Type{<:AbstractFockAddress{N}}) where {N} = N
+
+"""
+    num_modes(::Type{<:AbstractFockAddress})
+    num_modes(::AbstractFockAddress)
+
+Number of modes represented by address.
+"""
+num_modes(a::AbstractFockAddress) = num_modes(typeof(a))
+num_modes(::Type{<:AbstractFockAddress{<:Any,M}}) where {M} = M
+
+"""
+    num_components(::Type{<:AbstractFockAddress})
+    num_components(::AbstractFockAddress)
+
+Number of components in address.
+"""
+num_components(b::AbstractFockAddress) = num_components(typeof(b))
+
+"""
+    SingleComponentFockAddress{M}
+
+A type representing a single component Fock state with `M` modes.
+
+Implemented subtypes: [`BoseFS`](@ref), [`FermiFS`](@ref).
+
+# Supported functionality
+
+* [`find_mode`](@ref)
+* [`find_occupied_mode`](@ref)
+* [`num_occupied_modes`](@ref)
+* [`occupied_modes`](@ref)
+* [`move_particle`](@ref)
+
+"""
+abstract type SingleComponentFockAddress{N,M} <: AbstractFockAddress{N,M} end
+
+num_components(::Type{<:SingleComponentFockAddress}) = 1
+
+"""
+    find_mode(::SingleComponentFockAddress, i)
+
+Find the `i`-th mode in address. Returns [`BoseFSIndex`](@ref) for [`BoseFS`](@ref), and
+an integer for [`FermiFS`](@ref).
```

Comment:
```suggestion
Find the `i`-th mode in address. Returns [`BoseFSIndex`](@ref) for [`BoseFS`](@ref), and
an integer index for [`FermiFS`](@ref). Does not check bounds.
```
That is unless we want to add a `@boundscheck`, which might also make sense.